src/main/java/org/logstash/dissect/Dissector.java

```diff
@@ -53,34 +56,58 @@ public Dissector(String mapping) {
         Collections.sort(saveableFields, new FieldComparator());
     }
 
-    public int dissect(byte[] source, Map<String, Object> map) {
-        int pos = dissectValues(source);
-        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+    public int dissect(byte[] source, Map<String, Object> keyValueMap) {
+        ResolveValue resolveValue = resolve(source);
 
         for (Field field : saveableFields) {
-            field.append(map, resolver);
+            field.append(keyValueMap, resolveValue.resolver);
         }
-        return pos;
+        return resolveValue.position;
     }
 
     public int dissect(byte[] source, Event event) {
-        int pos = dissectValues(source);
-        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+        ResolveValue resolveValue = resolve(source);
 
         for (Field field : saveableFields) {
-            field.append(event, resolver);
+            field.append(event, resolveValue.resolver);
         }
-        return pos;
+        return resolveValue.position;
+    }
+
+    private ResolveValue resolve(byte[] source) {
+        final Map<Field, ValueRef> fieldValueRefMap = createFieldValueRefMap();
+        int pos = dissectValues(source, fieldValueRefMap);
+        ValueResolver resolver = new ValueResolver(source, fieldValueRefMap);
+        return new ResolveValue(pos, resolver);
     }
 
-    private int dissectValues(byte[] source) {
-        if (source.length == 0) return 0;
+    private class ResolveValue {
```

Comment:
That is easy to get wrong in Java. If you do it this way you actually get an implicit reference to the outer class which you don't need (and should avoid in general). Instead you should declare the class as `static`, i.e. `private static class ResolveValue`.